File: web-frontend/modules/core/utils/validator.js
import _ from 'lodash'

import { trueValues, falseValues } from '@baserow/modules/core/utils/constants'
import moment from '@baserow/modules/core/moment'
import { DateOnly, Timedelta } from '@baserow/modules/core/utils/date'
import { parseDurationString } from '@baserow/modules/core/utils/duration'

const isoRegex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{3})?Z?$/
const isoDateFormat = 'YYYY-MM-DD HH:mm:ss'
/**
 * Ensures that the value is a Numeral or can be converted to a numeric value.
 * @param {number|string} value - The value to ensure as a number.
 * @param allowNull {boolean} - Whether to allow null or empty values.
 * @returns {number|null} The value as a Number if conversion is successful.
 * @throws {Error} If the value is not a valid number or convertible to an number.
 */
export const ensureNumeric = (value, { allowNull = false } = {}) => {
  if (allowNull && (value === null || value === '' || value === undefined)) {
    return null
  }
  if (Number.isFinite(value)) {
    return value
  }
  if (typeof value === 'string' || value instanceof String) {
    if (/^([-+])?(\d+(\.\d+)?)$/.test(value)) {
      return Number(value)
    }
  }

  throw new Error(
    `Value '${value}' is not a valid number or convertible to a number.`
  )
}

/**
 * Ensures that the value is an integer or can be converted to an integer.
 * @param {number|string} value - The value to ensure as an integer.
 * @param {Object} options - Configuration options
 * @param {Boolean} [options.allowNegative=true] - Whether negative integer values
 * are allowed.
 * @returns {number|null} The value as an integer if conversion is successful, null otherwise.
 * @throws {Error} If the value is not a valid integer or convertible to an integer.
 */
export const ensureInteger = (value, { allowNegative = true } = {}) => {
  if (Number.isInteger(value)) {
    if (!allowNegative && value < 0) {
      throw new Error('Value is not a positive integer.')
    }
    return value
  }
  if (typeof value === 'string' || value instanceof String) {
    const integerRegex = allowNegative ? /^[+-]?\d+$/ : /^\d+$/
    if (integerRegex.test(value)) {
      const validInteger = Number(value)
      if (!allowNegative && validInteger < 0) {
        throw new Error('Value is not a positive integer.')
      }
      return validInteger
    }
  }
  if (value instanceof Timedelta) {
    const validInteger = Math.trunc(value.ms / 1000)
    if (!allowNegative && validInteger < 0) {
      throw new Error('Value is not a positive integer.')
    }
    return validInteger
  }
  throw new Error(
    `Value '${value}' is not a valid integer or convertible to an integer.`
  )
}

/**
 * Ensures that the value is a non-negative integer.
 * @param {*} value - The value to ensure is a non-negative integer.
 * @param {Object} options - Configuration options
 * @param {Boolean} [options.allowNull=false] - Whether to return null if value is null
 * @returns {number|null} The value as an integer if conversion is successful,
 * or null if allowNull is true or throwError is false
 * @throws {Error} If the value is not a valid non-negative integer
 */
export const ensurePositiveInteger = (value, { allowNull = false } = {}) => {
  if (allowNull && (value === null || value === '')) {
    return null
  }
  return ensureInteger(value, { allowNegative: false })
}
/**
 * Ensures that the value is a string or try to convert it.
 * @param {*} value - The value to ensure as a string.
 * @param {Boolean} allowEmpty - Whether we should throw an error if `value` is empty.
 * @returns {string} The value as a string.
 * @throws {Error} If !allowEmpty and the `value` is empty.
 */
export const ensureString = (value, { allowEmpty = true } = {}) => {
  if (
    value === null ||
    value === undefined ||
    value === '' ||
    (Array.isArray(value) && !value.length) ||
    (typeof value === 'object' && !(value instanceof Date) && _.isEmpty(value))
  ) {
    if (!allowEmpty) {
      throw new Error('A valid String is required.')
    }
    return ''
  }

  if (Array.isArray(value)) {
    // convert item into a string recursively
    const results = value.map((item) => ensureString(item))
    return results.join(',')
  } else if (value instanceof DateOnly) {
    return value.toString()
  } else if (value instanceof Date) {
    if (!isNaN(value)) {
      return moment(value).format(isoDateFormat)
    }
  } else if (value instanceof Timedelta) {
    return `${Math.floor(value.ms / 1000)}`
  } else if (typeof value === 'object') {
    // If it's a file we just extract the name
    if (value.__file__) {
      return value.name
    }
    return JSON.stringify(value)
  } else if (typeof value === 'string') {
    // Special case for iso strings. We parse them as dates for better display.
    if (isoRegex.test(value)) {
      const date = moment(value)
      if (date.isValid()) {
        return date.format(isoDateFormat)
      }
    }
  }
  return `${value}`
}

/**
 * Ensures the value is a string or an integer, otherwise tries to
 * convert the value to a string.
 *
 * @param {*} value - The value to ensure as a string or integer.
 * @param {Boolean} allowEmpty - Whether we should throw an error if `value` is empty.
 * @returns {string|number} The value as a string or integer.
 * @throws {Error} If !allowEmpty and the `value` is empty.
 */
export const ensureStringOrInteger = (value, { allowEmpty = true } = {}) => {
  if (Number.isInteger(value)) {
    return value
  }

  return ensureString(value, { allowEmpty })
}

/**
 * Ensure that the value is an array or try to convert it.
 * Strings will be treated as comma separated values.
 * Other data types will be transformed into a single element array.
 *
 * @param {*} value - The value to ensure as an array.
 * @param {Boolean} allowEmpty - Whether we should throw an error if `value` is empty.
 * @returns {any[]} The value as an array.
 * @throws {Error} if !allowEmpty and `value` is empty.
 */
export const ensureArray = (value, { allowEmpty = true } = {}) => {
  if (
    value === null ||
    value === undefined ||
    value === '' ||
    (Array.isArray(value) && !value.length)
  ) {
    if (!allowEmpty) {
      throw new Error('A non empty value is required.')
    }
    return []
  }
  let result
  if (Array.isArray(value)) {
    result = value
  } else if (typeof value === 'string') {
    result = value.split(',').map((item) => item.trim())
  } else {
    result = [value]
  }
  return result
}

/**
 * Ensures that the value is a non-empty string or try to convert it.
 * @param {*} value - The value to ensure as a string, which can't be blank.
 * @returns {string} The value as a string
 * @throws {Error} If `value` is empty.
 */
export const ensureNonEmptyString = (value, options) => {
  return ensureString(value, { ...options, allowEmpty: false })
}

/**
 * Ensures that the value is a boolean or convert it.
 * @param {*} value - The value to ensure as a boolean.
 * @param {boolean} useStrict - Whether to be strict in how the value is interpreted.
 * @returns {boolean} The value as a boolean.
 */
export const ensureBoolean = (value, { useStrict = true } = {}) => {
  if (trueValues.includes(value)) {
    return true
  } else if (falseValues.includes(value)) {
    return false
  }

  if (!useStrict) {
    return Boolean(value)
  }

  throw new Error('Value is not a valid boolean or convertible to a boolean.')
}

/**
 * Ensures that the value is a valid date or convert it.
 * @param {*} value - The value to ensure as a date
 * @returns {DateOnly} - The converted value as a DateOnly object.
 * @throws {Error} if `value` is not a valid date representation.
 */
export const ensureDate = (value, { allowEmpty = true } = {}) => {
  if (value === null || value === undefined || value === '') {
    if (!allowEmpty) {
      throw new Error('A non empty value is required.')
    }
    return null
  }

  const result = new DateOnly(value, 'YYYY-MM-DD', true)

  if (isNaN(result)) {
    throw new TypeError('Value is not a valid date or convertible to a date.')
  }

  return result
}

/**
 * Ensures that the value is a valid datetime or convert it.
 * @param {*} value - The value to ensure as a datetime
 * @param {boolean} allowEmpty - Whether to allow empty values.
 * @param {string} format - The format to use to parse the datetime.
 * @returns {Date} - The converted value as a Date object.
 * @throws {Error} if `value` is not a valid date representation.
 */
export const ensureDateTime = (
  value,
  { allowEmpty = true, format = moment.ISO_8601, useStrict = true } = {}
) => {
  if (value === null || value === undefined || value === '') {
    if (!allowEmpty) {
      throw new Error('A non empty value is required.')
    }
    return null
  }
  if (value instanceof Date) {
    return value
  } else {
    const parsed = format ? moment(value, format, useStrict) : moment(value)
    if (!parsed.isValid()) {
      throw new TypeError(
        'Value is not a valid datetime or convertible to a datetime.'
      )
    }
    return parsed.toDate()
  }
}

/**
 * Ensures that the value is a valid object or convert it.
 * @param {*} value - The value to ensure as an object
 * @returns {Object} - The converted value as an object.
 * @throws {Error} if `value` is not convertable to an object.
 */
export const ensureObject = (value) => {
  // Return early if it's an object
  if (value instanceof Object) {
    return value
  }

  if (value === null || value === undefined) {
    throw new TypeError(
      'Value is not a valid object or convertible to an object.'
    )
  }

  // If it's a string, try to parse it as JSON
  if (typeof value === 'string') {
    try {
      return JSON.parse(value)
    } catch {
      throw new TypeError('Value is not a valid JSON.')
    }
  }

  throw new TypeError(
    'Value is not a valid object or convertible to an object.'
  )
}

/**
 * Ensures that the value is a valid duration or converts it to one.
 * @param {*} value - The value to ensure as a duration.
 * @returns {Timedelta} - The converted value as a Timedelta.
 * @throws {TypeError} if `value` is not convertable to a Timedelta.
 */
export const ensureDuration = (value) => {
  if (value instanceof Timedelta) {
    return value
  }

  if (typeof value === 'string') {
    const result = parseDurationString(value)
    if (result !== null) return result

    try {
      return new Timedelta(ensureInteger(value) * 1000)
    } catch {
      throw new TypeError(
        `'${value}' is not a valid duration. Expected format: e.g. '1 day', '2 hours'.`
      )
    }
  }

  if (typeof value === 'number') {
    return new Timedelta(value * 1000)
  }

  throw new TypeError('Value cannot be converted to a duration.')
}

/**
 * Normalizes a value into a JSON-serializable counterpart, mirroring the
 * backend `ensure_json_serializable`. Special runtime types are converted to
 * the same representation the backend uses, so `to_json` produces identical
 * output on both sides: a `Timedelta` becomes its number of seconds. Arrays and
 * plain objects are normalized recursively; every other value is returned
 * unchanged so native JSON types pass through untouched.
 *
 * @param {*} value - The value to normalize.
 * @returns {*} The JSON-serializable value.
 */
export const ensureJsonSerializable = (value) => {
  if (value instanceof Timedelta) {
    return Math.floor(value.ms / 1000)
  }

  if (Array.isArray(value)) {
    return value.map((item) => ensureJsonSerializable(item))
  }

  if (_.isPlainObject(value)) {
    return _.mapValues(value, (item) => ensureJsonSerializable(item))
  }

  return value
}

/**
 * Decodes a JSON string if possible, otherwise returns the value unchanged,
 * mirroring the backend `ensure_deserialized_json`. Values that are not strings
 * are already deserialized and are returned as-is, even when `strict` is true.
 *
 * @param {*} value - The value to decode if it is a valid JSON string.
 * @param {Boolean} strict - If true, throw when `value` is a string that cannot
 *   be decoded as JSON, instead of returning it unchanged.
 * @returns {*} The decoded JSON value, or `value` when it is not a JSON string.
 * @throws {Error} If `strict` is true and `value` is a string that is not valid
 *   JSON.
 */
export const ensureDeserializedJson = (value, { strict = false } = {}) => {
  if (typeof value === 'string') {
    try {
      return JSON.parse(value)
    } catch (e) {
      if (strict) {
        throw new Error('Value is not valid JSON.', { cause: e })
      }
    }
  }

  return value
}

File: backend/src/baserow/core/formula/validator.py
import json
import re
from datetime import date, datetime, timedelta
from decimal import Decimal
from json.decoder import JSONDecodeError
from typing import Any, List, Optional, Union

from django.core.exceptions import ValidationError
from django.core.serializers.json import DjangoJSONEncoder
from django.core.validators import validate_email

from baserow.contrib.database.fields.constants import (
    BASEROW_BOOLEAN_FIELD_FALSE_VALUES,
    BASEROW_BOOLEAN_FIELD_TRUE_VALUES,
)
from baserow.core.datetime import FormattedDate, FormattedDateTime
from baserow.core.duration import parse_duration_string
from baserow.core.formula.service_file import ServiceFile
from baserow.core.models import UserFile
from baserow.core.utils import split_comma_separated_string


def ensure_file(value: Any) -> ServiceFile:
    if isinstance(value, ServiceFile):
        return value

    if isinstance(value, UserFile):
        return ServiceFile.from_user_file(value)

    if isinstance(value, dict):
        if value.get("__file__") or ("name" in value and "url" in value):
            return ServiceFile.from_serialized(value)

        raise ValidationError("A valid file or url is required.")

    if isinstance(value, str):
        return ServiceFile.from_remote_url(value)

    raise ValidationError("A valid file or url is required.")


def ensure_boolean(value: Any, strict=True) -> bool:
    """
    Ensures that the value is a boolean or converts it.

    :param value: The value to ensure as a boolean.
    :param strict: If not strict, an attempt is made to cast the value to a bool.
    :return: The value as a boolean.
    :raises ValidationError: if the value is not a valid boolean or convertible to a
        boolean.
    """

    if value in BASEROW_BOOLEAN_FIELD_TRUE_VALUES:
        return True
    elif value in BASEROW_BOOLEAN_FIELD_FALSE_VALUES:
        return False

    if not strict:
        return bool(value)

    raise ValidationError("Value is not a valid boolean or convertible to a boolean.")


def ensure_numeric(
    value: Any, allow_null: bool = False
) -> Optional[Union[int, float, Decimal]]:
    """
    Ensures that the value is a number or can be converted to a numeric value.

    :param value: The value to ensure as a number.
    :param allow_null: Whether to allow null or empty values.
    :return: The value as a number (int, float, or Decimal) if conversion is successful.
    :raises ValidationError: If the value is not a valid number or convertible
    to a number.
    """

    if allow_null and (value is None or value == ""):
        return None

    # Handle numeric types directly
    if isinstance(value, (int, float, Decimal)) and not isinstance(value, bool):
        return value

    # Handle string conversion
    if isinstance(value, str):
        # Check if the string matches a valid number pattern
        if re.match(r"^([-+])?(\d+(\.\d+)?)$", value):
            # Convert to int if it's a whole number, otherwise float
            try:
                num_value = float(value)
                if num_value.is_integer():
                    return int(value)
                return num_value
            except ValueError:
                # If float conversion fails, try Decimal as a fallback
                try:
                    return Decimal(value)
                except Exception as exc:
                    raise ValidationError(
                        f"Value '{value}' is not a valid number or convertible to a number."
                    ) from exc

    raise ValidationError(
        f"Value '{value}' is not a valid number or convertible to a number."
    )


def ensure_integer(
    value: Any, allow_empty: bool = False, allow_negative: bool = True
) -> Optional[int]:
    """
    Ensures that the value is an integer or can be converted to an integer.
    Raises a ValidationError if the value is not a valid integer or convertible to an
    integer.

    :param value: The value to ensure as an integer.
    :param allow_empty: Whether we should throw an error if `value` is empty.
    :param allow_negative: Whether negative integer values are allowed.
    :return: The value as an integer if conversion is successful.
    :raises ValidationError: If the value is not a valid integer or convertible to an
        integer.
    """

    if value is None or value == "":
        if not allow_empty:
            raise ValidationError("The value is required")
        return None

    if isinstance(value, bool):
        raise ValidationError(
            "The value must be an integer or convertible to an integer."
        )

    integer_pattern = r"^[+-]?\d+$" if allow_negative else r"^\d+$"
    if isinstance(value, str) and not re.match(integer_pattern, value):
        raise ValidationError(
            "The value must be an integer or convertible to an integer."
        )

    if isinstance(value, (float, Decimal)) and value != int(value):
        raise ValidationError(
            "The value must be an integer or convertible to an integer."
        )

    try:
        int_value = (
            int(value.total_seconds()) if isinstance(value, timedelta) else int(value)
        )
    except (ValueError, TypeError) as exc:
        raise ValidationError(
            "The value must be an integer or convertible to an integer."
        ) from exc

    if not allow_negative and int_value < 0:
        raise ValidationError(
            "The value must be a non-negative integer or convertible to one."
        )

    return int_value


def ensure_string(value: Any, allow_empty: bool = True) -> str:
    """
    Ensures that the value is a string or try to convert it.

    :param value: The value to ensure as a string.
    :param allow_empty: Whether we should throw an error if `value` is empty.
    :return: The value as a string.
    :raises ValidationError: If not allow_empty and the `value` is empty.
    """

    if value is None or value == "" or value == [] or value == {}:
        if not allow_empty:
            raise ValidationError("A valid String is required.")
        return ""

    if isinstance(value, bool):
        # To match the frontend
        return "true" if value else "false"

    if isinstance(value, timedelta):
        return str(int(value.total_seconds()))

    if isinstance(value, list):
        results = [ensure_string(item) for item in value if item]
        return ",".join(results)
    elif isinstance(value, dict):
        return json.dumps(value)

    return str(value)


def ensure_string_or_integer(value: Any, allow_empty: bool = True) -> Union[int, str]:
    """
    Ensures that the value is a string or an integer.

    :param value: The value to ensure as a string.
    :param allow_empty: Whether we should throw an error if `value` is empty.
    :return: The value as a string.
    :raises ValidationError: If not allow_empty and the `value` is empty.
    """

    if isinstance(value, int):
        return value

    return ensure_string(value, allow_empty=allow_empty)


def ensure_array(value: Any, allow_empty: bool = True) -> List[Any]:
    """
    Ensure that the value is an array or try to convert it.
    Strings will be treated as comma separated values.
    Other data types will be transformed into a single element array.

    :param value: The value to ensure as an array.
    :param allow_empty: Whether we should raise an error if `value` is empty.
    :return: The value as an array.
    :raises ValidationError: if not allow_empty and `value` is empty.
    """

    if value is None or value == "" or (isinstance(value, list) and not value):
        if not allow_empty:
            raise ValidationError("A non empty value is required.")
        return []

    if isinstance(value, list):
        return value

    if isinstance(value, str):
        try:
            return [item.strip() for item in split_comma_separated_string(value)]
        except ValueError as e:
            raise ValidationError(
                "The provided value is not a valid comma separated string."
            ) from e

    return [value]


def ensure_email(value: Any) -> str:
    """
    Ensures that the value is an email or can be converted to an email.
    :param value: The value to ensure as an email.
    :return: The value as an email.
    :raises ValidationError: If the value is not a valid email or convertible to an
      email.
    """

    str_value = ensure_string(value)

    validate_email(str_value)

    return str_value


def ensure_date(value: Any) -> Optional[date]:
    """
    Ensures that the value is a date or can be converted to a date.
    :param value: The value to ensure as a date.
    :return: The value as a date.
    :raises ValidationError: If the value is not a valid date or convertible to a date.
    """

    try:
        return FormattedDate(value).date if value is not None else None
    except (ValueError, TypeError) as exc:
        raise ValidationError("Value cannot be converted to a date.") from exc


def ensure_datetime(
    value: Any, date_format: str = "%Y-%m-%d %H:%M", strict: bool = False
) -> Optional[datetime]:
    """
    Ensures that the value is a datetime or can be converted to a datetime.
    :param value: The value to ensure as a datetime.
    :param date_format: The format to use to parse the datetime.
    :param strict: If True, always validates the date format.
    :return: The value as a datetime.
    :raises ValidationError: If the value is not a valid datetime or convertible to a
        datetime.
    """

    if value is None:
        return None

    try:
        if strict and isinstance(value, str):
            parsed = datetime.strptime(value, date_format)
            return FormattedDateTime(parsed, date_format=date_format).datetime
        return FormattedDateTime(value, date_format=date_format).datetime
    except (ValueError, TypeError) as exc:
        raise ValidationError("Value cannot be converted to a datetime.") from exc


def ensure_duration(value: Any) -> timedelta:
    """
    Ensures that the value is a timedelta or is convertible to a timedelta.
    :param value: The value to ensure as a timedelta.
    :return: The value as a timedelta.
    :raises ValidationError: If the value is not a valid timedelta or
        convertible to a timedelta.
    """

    if isinstance(value, timedelta):
        return value

    if isinstance(value, str):
        result = parse_duration_string(value)
        if result is not None:
            return result

        try:
            return timedelta(seconds=ensure_integer(value))
        except ValidationError:
            pass

        raise ValidationError(
            f"'{value}' is not a valid duration. "
            f"Expected format: e.g. '1 day', '2 hours'."
        )

    if isinstance(value, (int, float)):
        return timedelta(seconds=value)

    raise ValidationError("Value cannot be converted to a duration.")


def ensure_object(value: Any) -> Optional[dict]:
    """
    Ensures that the value is a dict or can be converted to a dict.
    :param value: The value to ensure as a dict.
    :return: The value as a dict.
    :raises ValidationError: If the value is not a valid dict or convertible to a
        dict.
    """

    if isinstance(value, dict):
        return value

    if isinstance(value, str):
        try:
            return json.loads(value)
        except (TypeError, JSONDecodeError) as exc:
            raise ValidationError("Value is not a JSON.") from exc

    raise ValidationError("Value cannot be converted to a dict.")


class BaserowFormulaJSONEncoder(DjangoJSONEncoder):
    def default(self, obj):
        if isinstance(obj, timedelta):
            return ensure_integer(obj)

        return super().default(obj)


def ensure_deserialized_json(value: Any, strict: bool = False) -> Any:
    """
    Decode a JSON string if possible, otherwise return the value unchanged.

    Values that are not strings are already deserialized and are returned as-is,
    even when `strict` is True.

    :param value: The value to decode if it is a valid JSON string.
    :param strict: If True, raise a ValidationError when `value` is a string that
        cannot be decoded as JSON, instead of returning it unchanged.
    :return: The decoded JSON value if `value` is a valid JSON string, otherwise
        `value`.
    :raises ValidationError: If `strict` is True and `value` is a string that is
        not valid JSON.
    """

    if isinstance(value, str):
        try:
            return json.loads(value)
        except (TypeError, JSONDecodeError) as exc:
            if strict:
                raise ValidationError("Value is not valid JSON.") from exc

    return value


def ensure_json_serializable(value: Any) -> Any:
    """
    Ensures that the value can be converted to a JSON value.
    Python types supported by Django's JSON encoder, like dates and datetimes, are
    converted to their JSON representation.

    :param value: The value to ensure as a JSON value.
    :return: The JSON-compatible value.
    :raises ValidationError: If the value cannot be converted to JSON.
    """

    try:
        return json.loads(
            json.dumps(value, cls=BaserowFormulaJSONEncoder, allow_nan=False)
        )
    except (TypeError, ValueError) as exc:
        raise ValidationError("Value cannot be converted to a JSON value.") from exc
